job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -272,6 +292,50 @@ private void copyLogs() {
             }
         }
     }
+
+    private String getLogPrefix(FlowNode node) {
+        BodyInvocationAction bodyInvocationAction = node.getAction(BodyInvocationAction.class);
+
+        if (bodyInvocationAction != null) {
+            BodyExecutionLabelAction bodyLabelAction = node.getAction(BodyExecutionLabelAction.class);
+
+            if (bodyLabelAction != null) {
+                return bodyLabelAction.getBodyLabel();
+            } else {
+                // This exec body does not have a body label. Stop now.
+                return null;
+            }
+        }
+
+        List<FlowNode> parents = node.getParents();
+        if (parents != null && !parents.isEmpty()) {
+            for (FlowNode parent : parents) {
+                String prefix = getLogPrefix(parent);
+                if (prefix != null) {
+                    return prefix;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    protected static final class LogLinePrefixOutputFilter extends LineTransformationOutputStream {
```

Comment:
If you do want to write tests for it—probably overkill since the hard work is done in the superclass, but who knows—then I would just give it default (“package-private”) access.